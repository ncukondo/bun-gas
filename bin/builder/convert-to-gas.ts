import { build as buildUsingVite } from "vite";
import fs from "fs";
import path from "path";
import { ensureDirSync } from "fs-extra";

const build = async (filename: string, name: string) => {
  console.log(filename);
  const res = await buildUsingVite({
    root: process.cwd(),
    build: {
      write: false,
      minify: false,
      lib: {
        entry: filename,
        name,
        formats: ["iife"],
      },
    },
  });
  const code = Array.isArray(res) ? res[0].output[0].code : "";
  return code;
};

const extractEntryExports = (code: string, globalName: string) => {
  const exports = new Function(`${code} return ${globalName}`)();
  const isFunction = (x: unknown): x is (...args: unknown[]) => unknown => typeof x === "function";
  return Object.entries(exports)
    .map(([name, obj]) => {
      return isFunction(obj)
        ? `function ${name}(...args){ return ${globalName}.${name}(...args);}`
        : `const ${name} = ${globalName}.${name};`;
    })
    .join("\n");
};

const convertToGoogleAppsScript = async (filename: string, globalName: string) => {
  const code = await build(filename, globalName);
  const exports = extractEntryExports(code, globalName);
  return `${code}\n\n${exports}\n`;
};

const convertAndWriteGoogleAppsScript = async (
  filename: string,
  globalName: string,
  output: string,
) => {
  const code = await convertToGoogleAppsScript(filename, globalName);
  const outputDir = path.dirname(output);
  ensureDirSync(outputDir);
  fs.writeFileSync(output, code);
};

export { convertToGoogleAppsScript, convertAndWriteGoogleAppsScript };
