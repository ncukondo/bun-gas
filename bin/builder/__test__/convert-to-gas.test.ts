import { describe, expect, test } from "bun:test";
import { convertToGoogleAppsScript } from "../convert-to-gas";
import path from "path";

describe("convert-to-gas", () => {
  const __dirname = path.dirname(new URL(import.meta.url).pathname);
  const sampleEntry = path.resolve(__dirname, "./sample/main.ts");
  test("convertToGoogleAppsScript", async () => {
    const code = await convertToGoogleAppsScript(sampleEntry, "main");
    expect(code).toContain("function sayHello");
    expect(code).toContain("const hello");
    expect(code).toMatchSnapshot();
  });
});
