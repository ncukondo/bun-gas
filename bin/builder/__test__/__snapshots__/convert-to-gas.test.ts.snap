// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`convert-to-gas convertToGoogleAppsScript 1`] = `
"var main = function(exports) {
  "use strict";
  const hello = "world";
  const sayHello = (name) => {
    console.log(\`Hello ${name}!\`);
  };
  exports.hello = hello;
  exports.sayHello = sayHello;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
  return exports;
}({});


const hello = main.hello;
function sayHello(...args){ return main.sayHello(...args);}
"
`;

exports[`convert-to-gas convertToGoogleAppsScript 1`] = `
"var main = function(exports) {
  "use strict";
  const hello = "world";
  const sayHello = (name) => {
    console.log(\`Hello ${name}!\`);
  };
  exports.hello = hello;
  exports.sayHello = sayHello;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
  return exports;
}({});


const hello = main.hello;
function sayHello(...args){ return main.sayHello(...args);}
"
`;

exports[`convert-to-gas convertToGoogleAppsScript 1`] = `
"var main = function(exports) {
  "use strict";
  const hello = "world";
  const sayHello = (name) => {
    console.log(\`Hello ${name}!\`);
  };
  exports.hello = hello;
  exports.sayHello = sayHello;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
  return exports;
}({});


const hello = main.hello;
function sayHello(...args){ return main.sayHello(...args);}
"
`;

exports[`convert-to-gas convertToGoogleAppsScript 1`] = `
"var main = function(exports) {
  "use strict";
  const hello = "world";
  const sayHello = (name) => {
    console.log(\`Hello ${name}!\`);
  };
  exports.hello = hello;
  exports.sayHello = sayHello;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
  return exports;
}({});


const hello = main.hello;
function sayHello(...args){ return main.sayHello(...args);}
"
`;

exports[`convert-to-gas convertToGoogleAppsScript 1`] = `
"var main = function(exports) {
  "use strict";
  const hello = "world";
  const sayHello = (name) => {
    console.log(\`Hello ${name}!\`);
  };
  exports.hello = hello;
  exports.sayHello = sayHello;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
  return exports;
}({});


const hello = main.hello;
function sayHello(...args){ return main.sayHello(...args);}
"
`;

exports[`convert-to-gas convertToGoogleAppsScript 1`] = `
"var main = function(exports) {
  "use strict";
  const hello = "world";
  const sayHello = (name) => {
    console.log(\`Hello ${name}!\`);
  };
  exports.hello = hello;
  exports.sayHello = sayHello;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
  return exports;
}({});


const hello = main.hello;
function sayHello(...args){ return main.sayHello(...args);}
"
`;
